arbor/include/arbor/cable_cell.hpp

```diff
@@ -29,14 +29,49 @@ struct lid_range {
         begin(b), end(e) {}
 };
 
-// Probe type for cell descriptions.
-struct cell_probe_address {
-    enum probe_kind {
-        membrane_voltage, membrane_current
-    };
+// Each kind of probe has its own type for representing its address,
+// as below:
 
+// Voltage estimate [mV] at `location`, possibly interpolated.
+struct cell_probe_membrane_voltage {
     mlocation location;
-    probe_kind kind;
+};
+
+// Total current density [A/m²] across membrane excluding capacitive current at `location`.
+struct cell_probe_total_ionic_current_density {
+    mlocation location;
+};
+
+// Value of state variable `state` in density mechanism `mechanism` in CV at `location`.
+struct cell_probe_density_state {
+    mlocation location;
+    std::string mechanism;
+    std::string state;
+};
+
+// Value of state variable `key` in point mechanism `source` at target `target`.
+struct cell_probe_point_state {
+    cell_lid_type target;
+    std::string mechanism;
+    std::string state;
+};
+
+// Current density [A/m²] across membrane attributed to the ion `source` at `location`.
```

Comment:
What about ion reversal potential? Would that also be handled as a separate type?
What about a `(location, ion, key)` tuple like we have for density mech. state variables?

Reply:
I prefer keeping the semantics in the type; the sort of data we can pluck from an ion say don't form an open class, so I would prefer not to switch on a value (i.e. enum kind).

Reply:
I want to exclude revpot, only because the whole precomputed reversal potential thing is a bit of a hack. We can come back to that later if there is a use case of course — adding it is not much work.